sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,550 @@
+    prev = &(*prev)->next;
+  }
+  *prev = s->wr_buf;
+  s->wr_buf = NULL;
+
+  // Attempt to send.
+  status_t res = OK_STATUS();
+  if (!s->sending) {
+    res = buffer_send(ctx, s);
+  }
+  return res;
+}
+
+// Indicates whether a stream is still busy, when attempting to finalize.
+static inline bool stream_busy(usb_logging_ctx_t *ctx,
+                               usb_logging_stream_t *s) {
+  if (s->wr_buf) {
+    buffer_completed(ctx, s);
+  }
+  return (s->send_buf != NULL);
+}
+
+// A buffer of logging data has been received by the host; we may try to
+// send another buffer if one is available.
+static status_t tx_done(void *s_v, usb_testutils_xfr_result_t result) {
+  usb_logging_stream_t *s = (usb_logging_stream_t *)s_v;
+  usb_logging_ctx_t *ctx = s->ctx;
+
+  s->sending = false;
+
+  // Return this buffer to the free list.
+  usb_logging_buf_t *buf = s->send_buf;
+  TRY_CHECK(buf != NULL);
+  s->send_buf = buf->next;
+  buffer_release(ctx, buf);
+
+  // Send the next, if there is one
+  if (s->send_buf) {
+    TRY(buffer_send(ctx, s));
+  }
+  return OK_STATUS();
+}
+
+// Callback function from the usb_testutils layer; presently this is called
+// every 16ms, provided that usb_testutils_poll is being called often.
+static status_t tx_flush(void *s_v) {
+  usb_logging_stream_t *s = (usb_logging_stream_t *)s_v;
+  usb_logging_ctx_t *ctx = s->ctx;
+
+  // We use this to prevent incomplete buffers of log data sitting here
+  // indefinitely and not reaching the host.
+  usb_logging_buf_t *buf = s->wr_buf;
+  if (buf && ibex_timeout_check(&s->flush_time)) {
+    if (buf->bytes_used > 0u) {
+      // Send everything that we have thus far collected in this buffer,
+      buffer_completed(ctx, s);
+    }
+    // Earliest time at which we shall again attempt flushing of log data on
+    // this stream.
+    s->flush_time = ibex_timeout_init(kUSBLogFlushInterval);
+  }
+
+  return OK_STATUS();
+}
+
+// Internal function for all USB logging.
+static status_t usb_logging_send(usb_logging_ctx_t *ctx,
+                                 usb_logging_stream_t *s, const uint8_t *data,
+                                 size_t len) {
+  size_t idx = 0U;
+  do {
+    // We must poll the testutils layer to keep things going...
+    TRY(usb_testutils_poll(ctx->usbutils));
+
+    // Try to ensure that we have a logging buffer.
+    if (!s->wr_buf) {
+      s->wr_buf = buffer_claim(ctx);
+    }
+    if (s->wr_buf) {
+      usb_logging_buf_t *buf = s->wr_buf;
+
+      // Add as much as we can to the current logging buffer.
+      size_t chunk = kUSBLogBufferSize - buf->bytes_used;
+      if (chunk > len - idx) {
+        chunk = len - idx;
+      }
+      if (s->remap) {
+        // Remap non-printable characters, to avoid issues with accidentally
+        // stimulated software flow control (XON/XOFF) at the host.
+        const uint8_t *sp = data + idx;
+        uint8_t *dp = &buf->data[buf->bytes_used];
+        for (unsigned i = 0u; i < chunk; ++i) {
```

Comment:
I would suggest avoiding `unsigned int` as the index. Use `size_t` instead.

`unsigned int` works poorly on RV64, and on RV32 ibex it has no difference from `size_t`.

Reply:
Changed, since none of these values is stored (repeatedly) in structures. uint16_t frankly is somewhat overkill for our system, since the available RAM is small...all of the indexes and buffer sizes are necessarily very limited too.

Reply:
In general `uint16_t` and `uint8_t` should be avoided for any arithmetic (okay for storage) because they don't compile to code as good as `uint32_t` or `uintptr_t`.